### PR TITLE
feat: Only submit applications that are in draft or returned status

### PR DESCRIPTION
**Pull request by inderps**

### CODE REVIEW

**Review by EmiPali — DISMISSED** [review]

File: src/main/java/io/github/genomicdatainfrastructure/daam/services/SubmitApplicationService.java [hunk 1]
```diff
@@ -5,38 +5,65 @@
...
-        this.remsApplicationCommandApi = applicationsApi;
+        this.remsApplicationCommandApi = applicationCommandApi;
+        this.remsApplicationsApi = applicationsApi;
     }
 
     public void submitApplication(String id) {
         var principal = (OidcJwtCallerPrincipal) identity.getPrincipal();
         String userId = principal.getClaim(USER_ID_CLAIM);
 
+        try {
```
Comment (brunopacheco1) [inline src/main/java/io/github/genomicdatainfrastructure/daam/services/SubmitApplicationService.java:48]: suggestion: to improve readability, you should move the try/catch that captures `40x` to another function, that retrieves the application if no `40x` happens.

File: src/main/java/io/github/genomicdatainfrastructure/daam/services/SubmitApplicationService.java [hunk 2]
```diff
@@ -5,38 +5,65 @@
...
+            var application = remsApplicationsApi.apiApplicationsApplicationIdGet(Long.valueOf(id), remsApiKey, userId);
+
+            if (!application.getApplicationApplicant().getUserid().equals(userId)) {
```
Comment (brunopacheco1) [inline src/main/java/io/github/genomicdatainfrastructure/daam/services/SubmitApplicationService.java:51]: suggestion: to improve readability, you should move all application content validation to another function.

File: src/main/java/io/github/genomicdatainfrastructure/daam/exceptions/ApplicationNotInCorrectStateException.java [hunk 3]
```diff
@@ -0,0 +1,7 @@
+package io.github.genomicdatainfrastructure.daam.exceptions;
+
+public class ApplicationNotInCorrectStateException extends RuntimeException {
+    public ApplicationNotInCorrectStateException(String message) {
+        super(message);
```
Comment (brunopacheco1) [inline src/main/java/io/github/genomicdatainfrastructure/daam/exceptions/ApplicationNotInCorrectStateException.java:5]: suggestion: to improve read,  the message is constant, you should move the text to this classe directly. And you should provide more meaningfull message, like the application id, the current application status and user id.

File: src/main/java/io/github/genomicdatainfrastructure/daam/exceptions/UserNotApplicantException.java [hunk 4]
```diff
@@ -0,0 +1,7 @@
+package io.github.genomicdatainfrastructure.daam.exceptions;
+
+public class UserNotApplicantException extends RuntimeException {
+    public UserNotApplicantException(String message) {
+        super(message);
```
Comment (brunopacheco1) [inline src/main/java/io/github/genomicdatainfrastructure/daam/exceptions/UserNotApplicantException.java:5]: suggestion: to improve read,  the message is constant, you should move the text to this classe directly. And you should provide more meaningfull message, like the application id and the user id.

File: src/main/java/io/github/genomicdatainfrastructure/daam/exceptions/ApplicationNotFoundException.java [hunk 5]
```diff
@@ -0,0 +1,7 @@
+package io.github.genomicdatainfrastructure.daam.exceptions;
+
+public class ApplicationNotFoundException extends RuntimeException {
+    public ApplicationNotFoundException(String message) {
+        super(message);
```
Comment (brunopacheco1) [inline src/main/java/io/github/genomicdatainfrastructure/daam/exceptions/ApplicationNotFoundException.java:5]: suggestion: to improve read,  the message is constant, you should move the text to this classe directly. And you should provide more meaningfull message, like the application id and the user id.

**Review by brunopacheco1 — CHANGES_REQUESTED** [review]

File: src/main/java/io/github/genomicdatainfrastructure/daam/services/SubmitApplicationService.java [hunk 6]
```diff
@@ -5,38 +5,65 @@
...
+            }
+
+            if (!application.getApplicationState().equals("application.state/draft") && !application.getApplicationState().equals("application.state/returned")) {
+                throw new ApplicationNotInCorrectStateException("Application is not in a submittable status");
+            }
+
+        } catch (WebApplicationException e) {
+            if (e.getResponse().getStatus() == 404) {
+                throw new ApplicationNotFoundException("Application not found");
+            } else {
```
Comment (brunopacheco1) [inline src/main/java/io/github/genomicdatainfrastructure/daam/services/SubmitApplicationService.java:62]: hint: this `else` is redundant, you could have simply done:

```
if (e.getResponse().getStatus() == 404) {
    throw new ApplicationNotFoundException("Application not found");
}
throw e;
```

File: src/main/java/io/github/genomicdatainfrastructure/daam/mappers/ApplicationNotInCorrectStateExceptionMapper.java [hunk 7]
```diff
@@ -0,0 +1,14 @@
...
+import io.github.genomicdatainfrastructure.daam.exceptions.ApplicationNotInCorrectStateException;
+import jakarta.ws.rs.core.Response;
+import jakarta.ws.rs.ext.ExceptionMapper;
+import jakarta.ws.rs.ext.Provider;
+
+@Provider
+public class ApplicationNotInCorrectStateExceptionMapper implements ExceptionMapper<ApplicationNotInCorrectStateException> {
+    @Override
+    public Response toResponse(ApplicationNotInCorrectStateException exception) {
+        return Response.status(Response.Status.BAD_REQUEST).entity(exception.getMessage()).build();
```
Comment (brunopacheco1) [inline src/main/java/io/github/genomicdatainfrastructure/daam/mappers/ApplicationNotInCorrectStateExceptionMapper.java:12]: important: `BAD_REQUEST` is misleading here, it should be either `CONFLICT - 409`, `UNPROCESSABLE_ENTITY - 422` or `PRECONDITION_REQUIRED - 428`. Perhaps the last one fits better.

File: src/main/openapi/rems.yaml [hunk 8]
```diff
@@ -70,6 +70,36 @@ paths:
...
+          schema:
+            type: integer
+            format: int64
+      responses:
+        '200':
+          description: ''
+          content:
+            application/json:
+              schema:
+                $ref: '#/components/schemas/ApplicationOverview'
```
[the remark this answers is not in the document]
Comment (inderps) [reply src/main/openapi/rems.yaml:102]: What do you mean by correct schema? ApplicationOverview seems to contain all application entity related properties

File: src/main/java/io/github/genomicdatainfrastructure/daam/services/SubmitApplicationService.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (inderps) [reply src/main/java/io/github/genomicdatainfrastructure/daam/services/SubmitApplicationService.java:52]: I looked at the documentaion, they only throw 404. No mention of 403

File: src/test/java/io/github/genomicdatainfrastructure/daam/api/ApplicationCommandApiImplTest.java [hunk 9]
```diff
@@ -52,6 +53,40 @@ void submitApplication_when_authenticated() {
...
+        .auth()
+        .oauth2(getAccessToken("alice"))
+        .when()
+        .post("/api/v1/applications/12345/submit")
+        .then()
+        .statusCode(403);
+  }
+
+  @Test
+  void submitApplication_when_not_applicant() {
```
[the remark this answers is not in the document]
Comment (inderps) [reply src/test/java/io/github/genomicdatainfrastructure/daam/api/ApplicationCommandApiImplTest.java:66]: The user does exist but not the applicant for this application. 

Regarding error codes, i mixed up somehow

File: src/main/openapi/rems.yaml (hunk 8 above)
[the remark this answers is not in the document]
Comment (brunopacheco1) [reply src/main/openapi/rems.yaml:102]: https://rems-test.rahtiapp.fi/swagger-ui/index.html#/applications/get_api_applications__application_id_

https://rems-test.rahtiapp.fi/swagger-ui/index.html#/applications/get_api_my_applications

Indeed they look similar, but I did not fully review the response model of the two API, they might differ from each other. If you've checked it thoroughly, it should be okay them. Please don't forget to align with @admy7 as he is implementing this endpoint.

File: src/main/java/io/github/genomicdatainfrastructure/daam/services/SubmitApplicationService.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (brunopacheco1) [reply src/main/java/io/github/genomicdatainfrastructure/daam/services/SubmitApplicationService.java:62]: https://github.com/GenomicDataInfrastructure/gdi-userportal-deployment you can run it locally if you want, or mimic  their demo environment request/response API call with Postman or RestClient, https://rems-demo.rahtiapp.fi/

File: src/main/openapi/rems.yaml (hunk 8 above)
[the remark this answers is not in the document]
Comment (admy7) [reply src/main/openapi/rems.yaml:102]: I have used the `Application` schema, the one mentioned in Rems' documentation for the endpoint /api/applications/{application-id}

File: src/main/java/io/github/genomicdatainfrastructure/daam/services/SubmitApplicationService.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (brunopacheco1) [reply src/main/java/io/github/genomicdatainfrastructure/daam/services/SubmitApplicationService.java:62]: I can support on any of these two approach.

File: src/main/java/io/github/genomicdatainfrastructure/daam/exceptions/ApplicationNotInCorrectStateException.java [hunk 10]
```diff
@@ -0,0 +1,10 @@
+// SPDX-FileCopyrightText: 2024 PNED G.I.E.
+//
+// SPDX-License-Identifier: Apache-2.0
+package io.github.genomicdatainfrastructure.daam.exceptions;
+
+public class ApplicationNotInCorrectStateException extends RuntimeException {
+    public ApplicationNotInCorrectStateException(Long id, String state) {
+        super("Application " + id + " is not in correct state: " + state);
```
Comment (brunopacheco1) [inline src/main/java/io/github/genomicdatainfrastructure/daam/exceptions/ApplicationNotInCorrectStateException.java:8]: hint: For the next time, all the messages in the exceptions could be constants. then you could use them as templates for interpolation.

In Java you can do it many ways, for me the simplest is using `.formatted()` functions.

```
private static final String MESSAGE = "Application %s is not in correct state: %s";
...
MESSAGE.formatted(id, state)
...

File: src/main/java/io/github/genomicdatainfrastructure/daam/mappers/ApplicationNotFoundExceptionMapper.java [hunk 11]
```diff
@@ -0,0 +1,13 @@
...
+import io.github.genomicdatainfrastructure.daam.exceptions.ApplicationNotFoundException;
+import jakarta.ws.rs.ext.Provider;
+import jakarta.ws.rs.core.Response;
+import jakarta.ws.rs.ext.ExceptionMapper;
+
+@Provider
+public class ApplicationNotFoundExceptionMapper implements ExceptionMapper<ApplicationNotFoundException> {
+    @Override
+    public Response toResponse(ApplicationNotFoundException exception) {
+        return Response.status(Response.Status.NOT_FOUND).entity(exception.getMessage()).build();
```
[the remark this answers is not in the document]
Comment (brunopacheco1) [reply src/main/java/io/github/genomicdatainfrastructure/daam/mappers/ApplicationNotFoundExceptionMapper.java]: Type should mean the same as title, the different is the audience. Type is machine readable - thus should never change - whereas title is human readable and can change overtime.

This is a tiny detail though. We can take care in another PR. Anyway, thanks a lot for taking care of exception handling so fast and so well.

File: src/test/java/io/github/genomicdatainfrastructure/daam/api/ApplicationCommandApiImplTest.java (hunk 9 above)
[the remark this answers is not in the document]
Comment (brunopacheco1) [reply src/test/java/io/github/genomicdatainfrastructure/daam/api/ApplicationCommandApiImplTest.java:66]: Ah, cool. Good to know there are other users to test.

**Review by brunopacheco1 — DISMISSED** [review]
Hey @inderps thanks a lot for you pacience and for promptly improving the code.

**Review by brunopacheco1 — APPROVED** [review]

**Pull request merged**